### PR TITLE
Add OpenCL rev (var_value) implementations of matrix multiply and sum

### PR DESCRIPTION
**Pull request by t4c1**
## Summary

Adds OpenCL reverse mode implementations of matrix multiply and sum (sum is required for easy testing a function that returns non-scalar value).

This also includes a small bugfix for kernel generator transpose.

## Tests
Both multiply and sum are tested.

## Side Effects
Comparing to existing rev matrix multiply this implementation leaves data in the OpenCL device between forward and reverse pass. This can be faster as it avoid copying, but a model that also uses other OpenCL functions could run out of device memory at smaller problem sizes.

## Release notes
Added OpenCL reverse mode implementations of matrix multiply and sum.

## Checklist

- [ ] Math issue #(issue number)

- [x] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: test/unit/math/opencl/rev/multiply_test.cpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+TEST(OpenCLMatrixMultiply, prim_rev_values_small) {
+  int N = 2;
+  int M = 3;
+  int K = 4;
+
+  Eigen::MatrixXd a(N, M);
+  a << 1, 2, 3, 4, 5, 6;
+  Eigen::MatrixXd b(M, K);
+  b << 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1;
+  stan::math::test::compare_cpu_opencl_prim_rev(matrix_multiply_functor, a, b);
```
Comment (rok-cesnovar) [inline test/unit/math/opencl/rev/multiply_test.cpp:19]: As far as I understand, the tests only tests multiply(matrix_cl, matrix_cl) and multply(var_value, var_value). We need to also test multply(matrix_cl, var_value) multply(var_value, matrix_cl).

**Review by rok-cesnovar — CHANGES_REQUESTED** [review]
Looks great. These OpenCL rev implementations are so clean.

I just have a few comments on testing.

File: stan/math/opencl/rev/multiply.hpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * @tparam T_b type of second expression
+ * @param a first expression
+ * @param b second expression
+ * @return Matrix product of given arguments
+ */
+template <
+    typename T_a, typename T_b,
+    require_all_nonscalar_prim_or_rev_kernel_expression_t<T_a, T_b>* = nullptr,
+    require_any_var_t<T_a, T_b>* = nullptr>
+inline auto operator*(const T_a& a, const T_b& b) {
```
Comment (SteveBronder) [inline stan/math/opencl/rev/multiply.hpp:29]: I think the compiler needs access to the `multiply()` function not `*`

**Review by SteveBronder — COMMENTED** [review]
Would it be possible to add the opencl version to the test suite? If we want the opencl version to be swappable with the eigen version I think it would be better/safer if they all had the same tests

File: stan/math/opencl/rev/multiply.hpp (hunk 2 above)
Comment (rok-cesnovar) [reply stan/math/opencl/rev/multiply.hpp:29]: Oh yeah, we need both of these.

**Comment by t4c1** [comment]
Every test is a part of a test suite. Do you mean could they use the ad testing framework? Hmm, that would require some extensions to the framework, as the OpenCL implementations do not have fwd implementations. Also they should be checked that they match the Eigen results. This would be a significant amount of work, but it might be worth doing. Not a part of this PR.

**Comment by SteveBronder** [comment]
Yes what I mean is [here](https://github.com/stan-dev/math/blob/develop/test/unit/math/test_ad.hpp#L1810) we could add a `test_matvar_opencl()`. We could either enforce that every function that accepts `var<Eigen>` also accepts `var<matrix_cl>` or have a bool template parameter that says whether or not a `var<matrix_cl>` version is available. I personally prefer the first choice as I think it will be worth it in the long run

**Comment by SteveBronder** [comment]
Those functions in the ad test suite just check that `var<matrix>` and `matrix<var>` return the same results so we don't need an fvar extension

**Comment by t4c1** [comment]
>I personally prefer the first choice as I think it will be worth it in the long run

I think that would force anyone who implements a var_value overload to also implement a GPU overload. We have a number of non-GPU devs and some functions that can not be easily implemented using kernel generator and would need new kernels. I think that would be a bad idea.

I am not that familiar with the ad testing framework. If you have an idea how it can be done it might be better if you do it.

**Comment by SteveBronder** [comment]
> I think that would force anyone who implements a var_value overload to also implement a GPU overload. We have a number of non-GPU devs and some functions that can not be easily implemented using kernel generator and would need new kernels. I think that would be a bad idea.

Yeah that's fair

> I am not that familiar with the ad testing framework. If you have an idea how it can be done it might be better if you do it.

Yeah I can take a crack at it I'll make a branch off of this one to work on it.

One Q, for `multiply(var_value<matrix_cl<double>> a, var_value<Eigen::MatrixXd>)` should the second argument get promoted / moved over to a `var_value<matrix_cl<double>>`?

**Comment by t4c1** [comment]
Nope. Overloads with mix of OpenCL types and Eigen/std vector types are not supported. It is up to compiler to move all the arguments to either OpenCL device or host in such case.

**Comment by rok-cesnovar** [comment]
Will approve once the tests pass. 

Gregor is preparing me a model that uses a poission GLM and a multiplication for the X argument, where X is a parameter so we can then see test the speedup when a successive multiply and GLM are both run on the GPU without transfers back in between.

**Comment by t4c1** [comment]
Extending the testing to all combinations of prim and rev arguments uncovered a number of bugs in GLMs, including one in Eigen implementation.

**Review by rok-cesnovar — APPROVED** [review]
This looks cool beans to me. Would also like  @SteveBronder to approve this one, given that you had some comments. Thanks!

**Review by SteveBronder — APPROVED** [review]
lgtm!

**Pull request merged**